### PR TITLE
Add lazy-seq/1 to support finite lazy sequences

### PR DESCRIPTION
**Pull request by profitware**
#289 Implement lazy sequences

`lazy-seq/1` takes a list to generate finite lazy sequence or another lazy sequence that passes as is. Lazy sequence is treated as finite if at any iteration it produces empty list instead of data as its head and nullary function for next iteration as its tail.

`drop/2` and `take/2` are modified to handle this behaviour. Commits for tests and documentation will follow today if anything is ok.

/CC @yurrriq @rvirding

### CODE REVIEW

File: src/clj.lfe [hunk 1]
```diff
@@ -643,7 +653,18 @@
...
 
+(defn- -lazy-seq
+  ([()] ())
+  ([`(,head . ,tail)] (cons head (fn [] (-lazy-seq tail)))))
+
+(defn- -drop
+  ([_ ()] ())
+  ([0 data] data)
+  ([n `(,_ . ,tail)]
+   (when (function? tail)) (-drop (dec n) (funcall tail))))
```
Comment (yurrriq) [inline src/clj.lfe]: I like to put guards on the same line as the pattern list when possible.

**Comment by yurrriq** [comment]
Before coffee this looks interesting. We'll want to make sure to have thorough test coverage so as not to let edge cases creep in.

File: src/clj.lfe [hunk 2]
```diff
@@ -515,6 +515,15 @@
   only computes subseqeuent values as needed."
   (fn [] (cons start (next func (funcall func start step) step))))
 
+(defn lazy-seq
+  "Return a lazy sequence (possibly infinite) from given lazy sequence `f`
+  or finite lazy sequence from given list `lst`. Lazy sequence is treated as
```
Comment (yurrriq) [inline src/clj.lfe]: :+1:

**Comment by rvirding** [comment]
My only question is that as this is going in the clj module is it "clojure compatible"? Or at least compatible enough?

File: src/clj.lfe (hunk 2 above)
Comment (yurrriq) [reply src/clj.lfe]: Upon further thought, I think this should be a macro, so as to delay evaluation. As it stands (unless I'm mistaken) `lst` is evaluated right away. @rvirding does `lambda` (`fn` is essentially an alias for `lambda`) delay evaluation too? Actually, I'm not sure that matters here as we're calling `lazy-seq` as a function so `lst` will be evaluated before applying `lazy-seq` to it. Let's make this a macro.

**Review by yurrriq — CHANGES_REQUESTED** [review]

**Comment by profitware** [comment]
`lazy-seq/1` here is compatible with `lazy-seq` in Clojure on how it is handled by `take` and `drop` and how it handles `list` and `lazy-seq` passed as an argument. As for overall compatibility the answer is no: the way `range/0`, `next/1` are implemented here (return nullary function) differs from Clojure's `range` and `iterate` (which are evaluated being once called).

**Comment by yurrriq** [comment]
I think the shape of the data in the current implementation is "compatible enough." (I said this elsewhere, I thought, but can't find it now), but I'd like to see it implemented as a macro, I'm nearly certain, so as to delay evaluation.

**Comment by rvirding** [comment]
`(lambda ... )` just returns a callable function so it doesn't do any evaluation. You have to call it with `funcall` to evaluate it. So that is not a reason to have a macro. And in the beginning you are just passing in a function which will be called for the first element.

Does next need to be, or should be, that "complex". Wouldn't it be enough to just return one element at a time? So the generator function takes no arguments and return either a list or () to show it is done. This would be conceptually simpler but maybe too simple.

I need to test it to get a fell for using it.

**Comment by yurrriq** [comment]
But if `lazy-seq` is a function (as it is now) doesn't this application first evaluate `lst`?

``` lfe
(lazy-seq lst)
```

**Comment by yurrriq** [comment]
I'd like to be able to do, e.g.

``` lfe
(lazy-seq (lists:map #'expensive-func/1 some-lat))
```

And have it only call `expensive-func/1` as I request elements of the resulting list via `take` and such.

**Comment by profitware** [comment]
Lists under `lazy-seq/1` macro are only evaluated when `take/2` or `drop/2` is called. Should anything else be done?

/CC @yurrriq @rvirding

**Comment by rvirding** [comment]
This now has a conflict in `clj.lfe`. Could you please rebase and fix that and I will pull it. Or do you want me to do it?

**Comment by profitware** [comment]
I've rebased my branch onto develop. Now I wonder if merging develop branch into mine would be preferred way to resolve conflicts next time this happens?

**Comment by yurrriq** [comment]
We usually rebase.

**Comment by profitware** [comment]
I think #278 and #289 may now be closed.

**Comment by rvirding** [comment]
I have found small bug in the code, I think. The pattern in the first clause of `lazy-seq` is

```
((()) ())
```

but should be

```
(`(()) ())
```

as it handles the case where the argument is a last containing the empty list. Is this correct? Or should be something else?

Found it because the compiler was complaining of "Warning: deprecated: no constructor in list pattern" because it is building a list containing `()` without a constructor. Could also write pattern as `(list ())`.

**Comment by profitware** [comment]
The pattern-matching statement cannot be written as 

```
(`(()) ())
```

because it doesn't match empty list given to `-lazy-seq` as the only argument: the whole macro is already escaped by the backquote. `(list ())` pattern doesn't work either (may be `(((list)) ())` will do the job?

I'm checking if the macro works by using the following two methods:

```
(slurp "src/clj.lfe")
(funcall (cdr (funcall (lazy-seq '(1)))))
```

in the LFE REPL should return `()` and running

```
make clean && make && rebar3 eunit
```

should not fail.

**Comment by rvirding** [comment]
The patterns assume lazy-seq only takes exactly one argument. If it is the empty list the it returns `()` otherwise it does stuff. Correct?

The second pattern `(,seq) expands to `(list seq)` a list of one argument. The second pattern `(())` does actually match against a list containing the empty list but it is missing a constructor, it should look like `(list ())` or `(()). It was once acceptable to not always have a list constructor in patterns but this is now deprecated and will soon not be allowed. This is why you get the deprecated warning when compiling the file.

My suggestion just added the constructor. (in 2 different ways)

**Comment by rvirding** [comment]
Just a trivial question: seeing it is a macro why have only one argument and not allow variable number of arguments? One reason could be that this could better allow you to build a list at runtime. I did not follow the discussion and am just curious.

**Comment by yurrriq** [comment]
The idea is to take a (possibly lazy) list and turn it into a lazy list, so the single argument makes sense to me. I'm not sure we can easily achieve the [behaviour I described above](https://github.com/rvirding/lfe/pull/290#issuecomment-254078650), so this seems good enough, for now anyway. I think we should definitely do this or similar, as @rvirding suggests:

``` diff
diff --git a/src/clj.lfe b/src/clj.lfe
index f0e62d7..d0dfd07 100644
--- a/src/clj.lfe
+++ b/src/clj.lfe
@@ -372,7 +372,7 @@
   or finite lazy sequence from given list `seq`. Lazy sequence is treated as
   finite if at any iteration it produces empty list instead of data as its
   head and nullary function for next iteration as its tail."
-  ((()) ())
+  (`(()) ())
   (`(,seq)
    `(lambda ()
       (let ((,'seq* ,seq))
```

**Comment by rvirding** [comment]
As I said I missed the discussion. Does it need to be a macro as it evaluates its argument?

I follow the "wisdom" of "don't use a macro when a function will do"

**Comment by yurrriq** [comment]
I think it needs to be a macro so it doesn't evaluate its argument. The more I think about this the more difficult it seems to properly mirror Clojure's `lazy-seq` behaviour. Instead I think we should rename (to avoid confusion) this macro, or remove it. If we want to support Clojure-inspired lazy lists (which we already do somewhat with `clj:{drop,take}/2`, `clj:cycle/1` et al.), then I think we'll need lazy mapping and folding functions. Without static types, it's pretty difficult to mimic Clojure's `lazy-seq`.

**Comment by yurrriq** [comment]
I think all this lazy list stuff needs to be rethought, actually.

``` lfe
> (clj:take 3 (clj:next (lambda (x step) (io:format "(+ ~p ~p)~n" (list x step)) (+ x step)) 1 1))
(+ 1 1)
(+ 2 1)
(+ 3 1)
(1 2 3)
```

Above you can see one extra value is computed and thrown away. I'm not sure when I'll get the time to give this a proper look, but I think we can do better.

**Comment by yurrriq** [comment]
I'm seeing these warnings as well, which pertain to the `lazy-seq` tests:

```
test/clj-tests.lfe:547: Warning: invalid function call
test/clj-tests.lfe:599: Warning: invalid function call
```

**Comment by rvirding** [comment]
I am trying to find out what causes these compilation warnings. Apparently that code is never called because at run-time it should result in a `badfun` error, which doesn't seem to occur (unless it is caught somewhere). I will try to find exactly what is generating these warnings.

The question is what we should do? Leave it in and try to fix it or reset it and fix it then re-pull? I am not a clojure person so I can't comment on whether it should be here.

**Comment by rvirding** [comment]
I tracked it down, and it is a sneaky one. There is no bug it is just the compiler optimiser which is a bit too smart. It propagates literal variable bindings, in this case binding `set*` to '(1 2 3), and then it works out that this is not a function so applying it with `(funcall set*)` will fail. It warns about this and replaces that with a call to erlang:error/1. However, it will never be called as the preceding test `erlang:is_function/2` will always fail.

**Comment by profitware** [comment]
I got it. I was wrongly thinking about `(`(()) ())` pattern in `-lazy-seq` inner fletrec. You are right about `lazy-seq` first pattern. Should I make a new pull-request or may you fix it by yourself?

**Comment by rvirding** [comment]
I will fix it.
